client-apis/streaming.md

```diff
@@ -0,0 +1,111 @@
+[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# Streaming
+
+## Version
+
+0.1.0
+
+## Changelog
+
+- 0.1.0
+
+  - [Clarify Streaming Encrypt and Decrypt](../changes/2020-07-06_clarify-streaming-encrypt-decrypt/change.md)
+
+## Overview
+
+The AWS Encryption SDK MAY provide APIs that enable streamed [encryption](encrypt.md)
+and [decryption](decrypt.md)
+Streaming is a framework for making bytes available to be processed
+by an operation sequentially and over time,
+and for outputting the result of that processing
+sequentially and over time.
+
+If an implementation requires holding the entire input in memory in order to perform the operation,
+that implementation SHOULD NOT provide an API that allows the caller to stream the operation.
+
+## Definitions
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC2119](https://tools.ietf.org/html/rfc2119).
+
+### Available Bytes
+
+In the scope of an operation, bytes are considered available if
+the operation has not yet processed those bytes
+and those bytes are intended to be processed,
+this intention being expressed through the specific streaming interface.
+
+For example, in a framework where a customer is requesting output bytes from an operation
+and that operation must read from some source in order to produce bytes,
+bytes read from the source are considered available and MUST be processed.
+Here it is the operation expressing intent to process a set of input bytes.
+Even if more bytes MAY be read from the source,
+bytes are not considered available to an operation
+until they have been consumed from the source.
+
+In a framework where a customer is sending input bytes to an operation
+and that operation must write those bytes to some sink,
```

Comment:
This is probably fine, but it *might* be worth either explaining a sink or using another word for it? It is fairly typical streaming terminology though, so up to you.

Reply:
Unless we can agree on a better word, I think "sink" here is sufficiently clear with it's intention here, especially when paired with "source" above.

Reply:
That's totally fine. Let's keep "sink".